xarray/core/indexes.py

```diff
@@ -45,72 +88,315 @@ def concat(
+        structure like a kd-tree that differs much from a simple array.
+
+        If not re-implemented in subclasses of Index, this method returns
+        ``None``, i.e., calling :py:meth:`Dataset.isel` will either drop the
+        index in the resulting dataset or pass it unchanged if its corresponding
+        coordinate(s) are not indexed.
+
+        Parameters
+        ----------
+        indexers : dict
+            A dictionary of positional indexers as passed from
+            :py:meth:`Dataset.isel` and where the entries have been filtered
+            for the current index.
+
+        Returns
+        -------
+        maybe_index : Index
+            A new Index object or ``None``.
+        """
         return None
 
     def sel(self, labels: dict[Any, Any]) -> IndexSelResult:
+        """Query the index with arbitrary coordinate label indexers.
+
+        Implementation is optional but required in order to support label-based
+        selection. Otherwise it will raise an error when trying to call
+        :py:meth:`Dataset.sel` with labels for this index coordinates.
+
+        Coordinate label indexers can be of many kinds, e.g., scalar, list,
+        tuple, array-like, slice, :py:class:`Variable`, :py:class:`DataArray`, etc.
+        It is the responsibility of the index to handle those indexers properly.
+
+        Parameters
+        ----------
+        labels : dict
+            A dictionary of coordinate label indexers passed from
+            :py:meth:`Dataset.sel` and where the entries have been filtered
+            for the current index.
+
+        Returns
+        -------
+        sel_results : :py:class:`IndexSelResult`
+            An index query result object that contains dimension positional indexers.
+            It may also contain new indexes, coordinate variables, etc.
+        """
         raise NotImplementedError(f"{self!r} doesn't support label-based selection")
 
-    def join(self: T_Index, other: T_Index, how: str = "inner") -> T_Index:
+    def join(self: T_Index, other: T_Index, how: JoinOptions = "inner") -> T_Index:
+        """Return a new index from the combination of this index with another
+        index of the same type.
+
+        Implementation is optional but required in order to support alignment.
+
+        Parameters
+        ----------
+        other : Index
+            The other Index object to combine with this index.
+        join : str, optional
+            Method for joining the two indexes (see :py:func:`~xarray.align`).
+
+        Returns
+        -------
+        joined : Index
+            A new Index object.
+        """
         raise NotImplementedError(
             f"{self!r} doesn't support alignment with inner/outer join method"
         )
 
     def reindex_like(self: T_Index, other: T_Index) -> dict[Hashable, Any]:
+        """Query the index with another index of the same type.
+
+        Implementation is optional but required in order to support alignment.
```

Comment:
Would be good to be specific about types of alignment here. Do we also need to mention reindexing?